{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "41aa083b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# defining string\n",
    "hcmv_fasta = \"\"\n",
    "#read in fasta file into a list\n",
    "with open('/Users/jengs/OneDrive - Oregon Health & Science University/From_Box/Pre-Share/PPG/scRNA_20250128/genomes_practice/HCMV.fa') as f:\n",
    "        hcmv_fasta_list= f.readlines()\n",
    "#remove first element as this is the seq name\n",
    "hcmv_fasta_list = hcmv_fasta_list[1:]\n",
    "#create one hcmv string\n",
    "#remove new line character\n",
    "for line in hcmv_fasta_list:\n",
    "        hcmv_fasta += line.rstrip()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "3e908ba9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#defining substring\n",
    "stop_codon = \"TAA\"\n",
    "stop_codon2 = \"TAG\"\n",
    "stop_codon3 = \"TGA\"\n",
    "forward_stop_codons=(stop_codon,stop_codon2,stop_codon3)\n",
    "polyA = \"AATAAA\"\n",
    "\n",
    "#reverse strand\n",
    "reverse_codon = \"TTA\"\n",
    "reverse_codon2 = \"CTA\"\n",
    "reverse_codon3 = \"TCA\"\n",
    "reverse_stop_codons=(reverse_codon,reverse_codon2,reverse_codon3)\n",
    "reverse_polyA = \"TTTATT\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "9c18fa39",
   "metadata": {},
   "outputs": [],
   "source": [
    "hcmv_gtf=pd.read_csv('/Users/jengs/OneDrive - Oregon Health & Science University/From_Box/Pre-Share/PPG/annotation_2025-04-09/annotation_2025_04_10_Meaghan.txt',sep=\"\\t\",header=None)\n",
    "hcmv_gtf.columns=[\"seq name\",\"source\",\"feature\",\"start\",\"end\",\"score\",\"strand\",\"frame\",\"attribute\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "0643174a",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "        with open('hcmv_meaghan_three_prime_utr.txt','w') as output:\n",
    "                #output.write(\"seq name\\tsource\\tfeature\\tstart\\tend\\tnew_start\\tnew_end\\tthreePrimeSeq\\tscore\\tstrand\\tframe\\tattribute\\n\")\n",
    "                for idx,gene_entry in hcmv_gtf.iterrows():\n",
    "                        strand=gene_entry[\"strand\"]\n",
    "                        #create a list of stop codon start locations within gene sequence\n",
    "                        stop_codon_start_list=[]\n",
    "                        #stop_codon_start_idx=0\n",
    "                        #create new start and new stop\n",
    "                        #new start will be start of 3 prime UTR for reverse strand\n",
    "                        #new stop will be end of 3 prime UTR for forward strand\n",
    "                        new_start=gene_entry[\"start\"]\n",
    "                        new_end=gene_entry[\"end\"]\n",
    "                        polyA_idx=0\n",
    "                        threePrimeSeq=\"\"\n",
    "                        #subset fasta with coordinates of transcript\n",
    "                        gene_fasta=hcmv_fasta[gene_entry[\"start\"]:gene_entry[\"end\"]]\n",
    "                        if strand==\"+\":\n",
    "                                for _, codon in enumerate(forward_stop_codons):\n",
    "                                        for m in re.finditer(codon, gene_fasta):\n",
    "                                                stop_codon_start_list.append(m.start(0))\n",
    "                                #get the max idx of start codon on forward strand\n",
    "                                if len(stop_codon_start_list)>0:\n",
    "                                    #search for poly A after stop codon\n",
    "                                    #add the start of transcript to the stop codon idx to get idx of hcmv genome\n",
    "                                    #max search at 500 bp\n",
    "                                    stop_codon_start_idx=gene_entry[\"start\"]+max(stop_codon_start_list)\n",
    "                                    #polyA_idx will be the new end coordinate\n",
    "                                    polyA_idx=re.search(polyA,hcmv_fasta[stop_codon_start_idx:(stop_codon_start_idx+500)])\n",
    "                                    if polyA_idx:\n",
    "                                        polyA_idx=stop_codon_start_idx+polyA_idx.end(0)\n",
    "                                    else:\n",
    "                                        polyA_idx=stop_codon_start_idx+500\n",
    "                                    new_end=polyA_idx\n",
    "                                    threePrimeSeq=hcmv_fasta[stop_codon_start_idx:polyA_idx]\n",
    "    \n",
    "                        else:\n",
    "                                rev_polyA_idx=[]\n",
    "                                for _, rev_codon in enumerate(reverse_stop_codons):\n",
    "                                    for m in re.finditer(rev_codon, gene_fasta):\n",
    "                                            stop_codon_start_list.append(m.end(0))\n",
    "                                #get minimum idx of start codon on reverse strand\n",
    "                                #search for revPolyA before stop codon\n",
    "                                #add start of transcript to stop codon idx to get idx of hcmv genome\n",
    "                                #max search at 500 bp\n",
    "                                #only change new_start if found stop codon\n",
    "                                if len(stop_codon_start_list)>0:\n",
    "                                    stop_codon_start_idx=gene_entry[\"start\"]+min(stop_codon_start_list)\n",
    "                                    for revPolyA in re.finditer(reverse_polyA,hcmv_fasta[stop_codon_start_idx-500:stop_codon_start_idx]):\n",
    "                                        rev_polyA_idx.append(revPolyA.start(0))\n",
    "                                    if len(rev_polyA_idx)>0:\n",
    "                                        #get the revpolyA closest to the stop codon\n",
    "                                        polyA_idx=stop_codon_start_idx-500+max(rev_polyA_idx)\n",
    "                                        new_start=polyA_idx\n",
    "                                    else:\n",
    "                                        polyA_idx=stop_codon_start_idx-500\n",
    "                                        new_start=polyA_idx\n",
    "                                        \n",
    "                                    \n",
    "                                threePrimeSeq=hcmv_fasta[new_start:stop_codon_start_idx]\n",
    "                        #output new line of gtf with putative 3 prime utr\n",
    "                        #commenting out extraneous cols used for double checking\n",
    "                        output.write(gene_entry[\"seq name\"] +\"\\t\")\n",
    "                        output.write(gene_entry[\"source\"] +\"\\t\")\n",
    "                        output.write(gene_entry[\"feature\"] +\"\\t\")\n",
    "                        #output.write(str(gene_entry[\"start\"]) +\"\\t\")\n",
    "                        #output.write(str(gene_entry[\"end\"]) +\"\\t\")\n",
    "                        output.write(str(new_start)+\"\\t\")\n",
    "                        output.write(str(new_end)+\"\\t\")\n",
    "                        #output.write(threePrimeSeq+\"\\t\")\n",
    "                        output.write(gene_entry[\"score\"] +\"\\t\")\n",
    "                        output.write(gene_entry[\"strand\"] +\"\\t\")\n",
    "                        output.write(gene_entry[\"frame\"] +\"\\t\")\n",
    "                        output.write(gene_entry[\"attribute\"] +\"\\n\")                        \n",
    "                \n",
    "        output.close()\n",
    "except IOError:\n",
    "        print(\"Error\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7a86fe3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ccc4476",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
